### PR TITLE
Ghost Float Ability

### PR DESCRIPTION
**Pull request by michaelbolt**
Changes character to ghost and adds ability to "Float" for several seconds before falling.

### CODE REVIEW

File: Assets/Scripts/Input/InputActions.cs
Comment (michaelbolt) [inline Assets/Scripts/Input/InputActions.cs:1]: this file adds a new `jump` action

File: Assets/Scripts/Input/InputActionsReader.cs [hunk 1]
```diff
@@ -17,6 +17,7 @@ public class InputActionsReader : MonoBehaviour
     public Vector2GameEvent gameplayLookEvent;
     public VoidGameEvent gameplayInteractPressed;
     public VoidGameEvent gameplayInteractReleased;
+    public BooleanGameEvent gameplayJumpPressed;
```
Comment (michaelbolt) [inline Assets/Scripts/Input/InputActionsReader.cs:20]: A new `GameEvent` to raise `true`/`false` when the `jump` action is pressed/released! 

The `BooleanGameEvent` itself is `Assets/ScriptableObjects/GameEvents/Input/GameplayJumpEventSO.asset`.

File: Assets/ScriptableObjects/GameEvents/Input/GameplayJumpEventSO.asset
Comment (michaelbolt) [inline Assets/ScriptableObjects/GameEvents/Input/GameplayJumpEventSO.asset:1]: This is the new `GameEvent` that gets raised when the `jump` button is pressed / released; anything can subscribe to it with a `GameEventListener` componnet

File: Assets/Scripts/Player/GhostModelMover.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
+using UnityEngine;
+
+/// <summary>This Component causes the model to bob up and down (like a ghost!).</summary>
+public class GhostModelMover : MonoBehaviour
+{
+    [Tooltip("Number of bobs to make in one second")]
+    public float bobFrequency = 0.7f;
+    [Tooltip("Maximum vertical height of bobbing")]
+    public float bobHeight = 0.1f;
+    [Tooltip("Vertical offset to apply to bobbing")]
```
Comment (michaelbolt) [inline Assets/Scripts/Player/GhostModelMover.cs]: should remove this, the `bobOffset` variable is unused and it makes for an unnecessary tooltip!

File: Assets/Scripts/Utilities/MyUtils.cs
Comment (michaelbolt) [inline Assets/Scripts/Utilities/MyUtils.cs:1]: this code now lives in `Assets/Scripts/Player/PlayerHorizontalMotion.cs`, as it was the only place it was being used.

File: Assets/Scripts/Player/PlayerHorizontalMotion.cs
Comment (michaelbolt) [inline Assets/Scripts/Player/PlayerHorizontalMotion.cs:1]: decoupled all horizontal motion from unity `Monobehavior` logic - so this is a standalone class now!

File: Assets/Scripts/Input/InputActions.cs
Comment (calebh94) [reply Assets/Scripts/Input/InputActions.cs:1]: Successfully tested on Mac with Switch Pro Controller 👍 👻

File: Assets/Scripts/Input/InputActionsReader.cs (hunk 1 above)
Comment (calebh94) [reply Assets/Scripts/Input/InputActionsReader.cs:20]: Thanks for documenting, will be useful.

File: Assets/Scripts/Player/GhostModelMover.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    private float modelOffset = 0.0f;
+
+    // Update is called once per frame
+    void Update()
+    {
+        float oldModelOffset = modelOffset;
+        modelOffset = Mathf.Cos(2.0f * Mathf.PI * bobFrequency * Time.time) * bobHeight;
+        transform.position = new Vector3(transform.position.x, transform.position.y - oldModelOffset + modelOffset, transform.position.z);
+    }
+}
```
Comment (calebh94) [inline Assets/Scripts/Player/GhostModelMover.cs:20]: I see, you are transforming the model each Update call. Is this not a feature that could be set as an Animation? The default animation could be the bobbing, and the ghost could have other animations when floating or jumping or talking.

File: Assets/Scripts/Player/PlayerVerticalMovement.cs [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+                // can release button to fall faster
+                if (!jumpPressed) { nextState = JumpState.RELEASED; }
+                // transition to FLOATING at peak of jump 
+                else if (verticalVelocity <= 0.0f) { nextState = JumpState.FLOATING; }
+                break;
+
+            case JumpState.FLOATING:
+                // can release button to fall faster
+                // OR exceed maxHovertime
+                if (!jumpPressed || (Time.time - currentStateStartTime) > maxHoverTime) { nextState = JumpState.RELEASED; }
```
Comment (calebh94) [inline Assets/Scripts/Player/PlayerVerticalMovement.cs:72]: Rather than using a Max Hover Time, what about we increase the speed at which you fall over time? Then items will reduce that value and eventuallly you can hover indefinitely, aka fly 😛

Comment (calebh94) [reply Assets/Scripts/Player/PlayerVerticalMovement.cs:72]: We could also give it the feel of flying, similar to A Short Hike.

Also, I see below where you do have the gravity during the jump. We can play with this more moving forward

**Review by calebh94 — APPROVED** [review]
The Ghost Float ability and the new Input and Interaction Systems are great. This is looking really good so far. Excited for the next steps with Interaction and Abilities.

File: Assets/Scripts/Player/GhostModelMover.cs (hunk 3 above)
Comment (michaelbolt) [reply Assets/Scripts/Player/GhostModelMover.cs:20]: yup! This could also be done in an animation - but this took all of six minutes to make and didn't require a bunch of animator state machine logic 🤷 

I'm also thinking it would be easier to separate this bobbing (which we can control in code very easily) from actually _animating_ the  ghost for different poses and emotions and things. When I've animated rigged characters, the less you have to pose them the better - and remembering to add a rhythmic bobbing will just make it a bit harder to do consistently.

So in short: yeah, I just faked an animation with #math lol

File: Assets/Scripts/Player/PlayerVerticalMovement.cs (hunk 4 above)
Comment (michaelbolt) [reply Assets/Scripts/Player/PlayerVerticalMovement.cs:72]: we could take a fall-speed based approach to it too - I just designed it this way because it made it easier for me to think about "how far" you can float for level design.

Like, if I know my max horizontal speed is 10 units/sec and my float time is 3 seconds then I know I can design a gap 30 units wide before the player can't cross it. But if I had to design in terms of acceleration and gravity values, I think it would get pretty wonky pretty quick! 

We could also have items increase the hover time - adding a second or two to that number is pretty ez

File: Assets/Scripts/Player/GhostModelMover.cs (hunk 2 above)
Comment (michaelbolt) [reply Assets/Scripts/Player/GhostModelMover.cs]: just removed this in 0cb7714f8552a4ae5fa499e61701ceb0269417b9
